gensim/similarities/levenshtein.py

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2018 Vit Novotny <witiko@mail.muni.cz>
+# Licensed under the GNU LGPL v2.1 - http://www.gnu.org/licenses/lgpl.html
+
+"""
+This module provides a namespace for functions that use the Levenshtein distance.
+"""
+
+import logging
+
+# If python-Levenshtein is available, import it.
+# If python-Levenshtein is unavailable, ImportError will be raised in levsim.
+try:
+    from Levenshtein import distance
+    LEVENSHTEIN_EXT = True
+except ImportError:
+    LEVENSHTEIN_EXT = False
+
+from gensim.similarities.termsim import TermSimilarityIndex
+
+logger = logging.getLogger(__name__)
+
+
+def levsim(t1, t2, alpha=1.8, beta=5.0):
```

Comment:
The method signature (even if not implemented/used at the moment) should contain a "max_distance" threshold. With the semantics that values above this threshold are uninteresting to the upstream application, computation can terminate early.

Such "early out" optimization is super cheap and makes a huge difference in practical applications, in my experience.

Reply:
I added the `max_distance` parameter in commit ac60615. Currently, [the Levenshtein distance library I use](https://pypi.org/project/python-Levenshtein/) does not support early termination, but it should be a straightforward future addition ([`_levenshtein.c`](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c) at lines [2304](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c#L2304), [2332](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c#L2332), [2352](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c#L2352), [2448](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c#L2448), [2483](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c#L2483), [2511](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c#L2511), [2531](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c#L2531), [3223](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c#L3223), [3856](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c#L3856), [4815](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c#L4815), [4935](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c#L4935), [5810](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c#L5810), and [5999](https://github.com/ztane/python-Levenshtein/blob/master/Levenshtein/_levenshtein.c#L5999)).

Reply:
I added the above notes to the future work section of the original post.

Reply:
Since there does not seem to be any rush to merge this, I started with the [modifications of the Levenshtein library](https://github.com/Witiko/python-Levenshtein/commit/0f9de6848d3bb5c9cde23ee05f8683e5563e7578). The interface, and the basic implementation is in place. Currently, the code only terminates prematurely, when `abs(len(t1), len(t2)) > max_distance`. There are likely other opportunities for early termination.

Reply:
In the light of this, I suggest we revert / forcibly roll back the last couple of commits (ac60615, 5154569, 729d185, 155dc58, and 7e52ef8) introducing `max_distance`, and `min_similarity`